File: ML_library/websocketserver.py
import asyncio
import websockets
import datetime
from pymongo import MongoClient
import os
from configure import config
from urllib.parse import unquote

mongo_client = MongoClient(os.getenv('MONGODB_HOST', 'localhost'))
db = mongo_client[config['mongo']['dbName']]

async def connect(websocket, path):
    # If the path is too short, abort
    if len(path) < len("/api/video/"):
        websocket.close()
        return

    # Split url into parts
    urlParts = path[len("/api/video/"):].split('/')
    # So the full URL starts with /api/video, but we're ignoring that.

    # If the path is still too short, abort
    if len(urlParts) < 4:
        websocket.close()
        return

    # Remove the user's ID from the URL
    user_id = urlParts.pop(0)
    # Remove the session cookie from the URL
    session_cookie = urlParts.pop(0)
    # Check if this user actually exists by attempting to get their _id. If they don't, abort
    # TODO: update this to work with the new sessions collection
    try:
        db.users.find_one({"_id": user_id, "sessionCookie": unquote(session_cookie)})["_id"]
    except:
        websocket.close()
        return


    # Now what's left in urlParts is e.g. /groupID/projectID/lectureID/extension
    # Or, you can omit lectureID and just have /groupID/projectID/extension
    # Or, you can omit lectureID and projectID and just have /groupID/extension
    # extension will be mp4/mp3/avi/mov/etc.
    # The IDs (like groupID) will be some string like "abcdefghi1234567"

    extension = urlParts[len(urlParts) - 1]

    # Obtain the current time
    now = datetime.datetime.now()

    # JSON to put into mongo later
    putInDB = {
        "extension": extension,
        "uploadedBy": user_id,
        "uploadedAt": now
    }

    # Depending on length of the parts, different ID will be used
    if len(urlParts) == 4:
        putInDB["lectureID"] = urlParts[2]
    elif len(urlParts) == 3:
        putInDB["projectID"] = urlParts[1]
    else:
        putInDB["groupID"] = urlParts[0]

    # Connect with mongo and insert the previous JSON object into the videos collection
    result = db['videos'].insert_one(putInDB)
    videoID = result.inserted_id

    # Send the videoID that was just recorded
    await websocket.send(str(videoID))

    # Path to save the video
    path = config['uploadsDir'] + '/' + str(videoID) + '.' + extension
    # TODO:
    # receive data, then append it to the file at path
    while True:
        try:
            data = await websocket.recv()
        except Exception as e:
            print("Ignoring error", e)
            try:
                websocket.close()
            finally:
                break
        else:
            f = open(path, 'a+')
            f.write(data)
        # TODO: Send data receive through extract frame/audio function


def start_server():
    asyncio.set_event_loop(asyncio.new_event_loop())
    start_server = websockets.serve(connect, '0.0.0.0', 5678)
    asyncio.get_event_loop().run_until_complete(start_server)
    asyncio.get_event_loop().run_forever()

File: backend/websocketserver.js
const WebSocket = require('ws');
const fs = require('fs');
const path = require('path');
const { Readable } = require('stream');
const cookieParser = require('cookie-parser')();
const mustBeLoggedIn = require('./api/must-be-logged-in.js');
const extractAudio = require('./processing/extract-audio.js');
const extractFrames = require('./processing/extract-frames.js');

// Accepts a WebSocket server (wss) and then sets it up
module.exports = wss => {
  wss.on('connection', async (ws, req) => {
    const urlParts = req.url
      .substr('/api/video/'.length)
      .split('/')
      .map(decodeURIComponent);
    // There have to be at least 4 parts (userID, sessionCookie, groupID, extension)
    if (urlParts.length < 4) {
      ws.close();
      return;
    }

    req.cookies = {};
    req.cookies.id = urlParts.shift();
    req.cookies.sessionCookie = urlParts.shift();

    // Make sure the user is logged in
    try {
      await mustBeLoggedIn.withoutResponse(req);
    } catch (err) {
      console.log(err);
      ws.close();
      return;
    }

    const extension = urlParts.pop(); // The very last thing is the extension
    // TODO: check the extension of the file here!

    let putInDB = {
      extension,
      uploadedBy: req.user._id.toString(),
      uploadedAt: new Date()
    };

    let [groupID, projectID, lectureID] = urlParts;
    if (lectureID) {
      putInDB.lectureID = lectureID;
    } else if (projectID) {
      putInDB.projectID = projectID;
    } else {
      putInDB.groupID = groupID;
    }

    const result = await mongo.collection('videos').insertOne(putInDB);

    // Send back the ID of the video that was just recorded
    const videoID = result.insertedId;
    ws.send(JSON.stringify({ videoID }));

    // Create a stream to save the video to a file
    const path = config.uploadsDir + '/' + videoID + '.' + extension;
    const videoStream = fs.createWriteStream(path);

    // These will be streams (or fake streams) used by extractFrames or extractAudio
    let videoStreamForFrames, videoStreamForAudio;

    // Only use a real stream if the server is configured for it
    // TODO: also check if the task scheduler will allow this processing
    if (config.processing.frames) {
      videoStreamForFrames = new Readable();
      videoStreamForFrames._read = () => {}; // Needed for some versions of Node.js
      extractFrames(videoStreamForFrames, videoID);
    } else {
      // Supply a fake stream instead of actually streaming to extract frames
      videoStreamForFrames = { push: () => {} };
    }

    if (config.processing.audio) {
      videoStreamForAudio = new Readable();
      videoStreamForAudio._read = () => {}; // Needed for some versions of Node.js
      extractAudio(videoStreamForAudio, videoID);
    } else {
      // Supply a fake stream instead of actually streaming to extract frames
      videoStreamForAudio = { push: () => {} };
    }

    ws.on('message', data => {
      videoStream.write(data);
      videoStreamForFrames.push(data);
      videoStreamForAudio.push(data);
    });
    ws.on('close', () => {
      videoStream.end();
      videoStreamForFrames.push(null);
      videoStreamForAudio.push(null);
    });
  });
};
